src/queue.rs

```diff
@@ -0,0 +1,163 @@
+mod job;
+pub use crate::entity::queue::{JobResult, JobStatus};
+pub use job::*;
+
+use crate::{
+    entity::queue::{ActiveModel, Entity, Model},
+    ApiConfig, Db,
+};
+use sea_orm::{ActiveModelTrait, DbErr, IntoActiveModel, Set, TransactionTrait};
+use std::{ops::Range, sync::Arc, time::Duration};
+use time::OffsetDateTime;
+use tokio::{
+    task::{JoinHandle, JoinSet},
+    time::sleep,
+};
+use trillium_tokio::{CloneCounterObserver, Stopper};
+
+#[derive(Clone, Debug)]
+pub struct Queue {
+    observer: CloneCounterObserver,
+    stopper: Stopper,
+    db: Db,
+    job_state: Arc<SharedJobState>,
+}
+/*
+These configuration variables may eventually be useful to put on ApiConfig
+*/
+const MAX_RETRY: i32 = 5;
+const QUEUE_CHECK_INTERVAL: Range<u64> = 10..20;
+const QUEUE_WORKER_COUNT: u8 = 2;
+
+fn reschedule_based_on_failure_count(failure_count: i32) -> Option<OffsetDateTime> {
+    if failure_count >= MAX_RETRY {
+        None
+    } else {
+        Some(
+            OffsetDateTime::now_utc()
+                + Duration::from_millis(
+                    1000 * 4_u64.pow(failure_count.try_into().unwrap()) + fastrand::u64(0..15000),
+                ),
+        )
+    }
+}
+
+impl Queue {
+    pub fn new(db: &Db, config: &ApiConfig) -> Self {
+        Self {
+            observer: Default::default(),
+            db: db.clone(),
+            stopper: Default::default(),
+            job_state: Arc::new(config.into()),
+        }
+    }
+
+    pub fn with_observer(mut self, observer: CloneCounterObserver) -> Self {
+        self.observer = observer;
+        self
+    }
+
+    pub fn with_stopper(mut self, stopper: Stopper) -> Self {
+        self.stopper = stopper;
+        self
+    }
+
+    pub async fn perform_one_queue_job(&self) -> Result<Option<Model>, DbErr> {
+        let tx = self.db.begin().await?;
+        let model = if let Some(queue_item) = Entity::next(&tx).await? {
+            let _counter = self.observer.counter();
+            let mut queue_item = queue_item.into_active_model();
+            let mut job = queue_item
+                .job
+                .take()
+                .expect("queue jobs should always have a Job");
```

Comment:
I think it's better to return an error here instead of crashing, especially since the method already returns `Result`. I suspect we'd just crash farther up the call stack, but maybe at some point we'll want to do something like emit a metric on fatal errors or exit with a particular status code and that's easier to do if we don't abort from all over the program.

Reply:
resolved in 9f4fcc8d006b8373cbf42b428412b8879daadb01